[ios] Place Page for the Track Recording

kirylkaveryn opened the pull request:
This PR implements the basic Track Recording Place Page (TR PP) screen.

Key features:
1. It matches the Track's PP
2. The track lightweight statistics is always in an up-to-date state
3. The track elevation profile data and statistics are updated continuously only when:
 - the PP screen is `Open` 
 - the application is in the `Active` state
This means that energy-inefficient elevation profile calculations and drawings will not be performed when the PP is not presented on the device screen. If the user selects some POI/bookmark the ele profile calculation will also **not** be performed.
4. Tapping on the `Start track recording` or on the `Right-top activity button` will open the TR PP. 
5. When the TR PP is opened the `Right-top activity button` will be hidden to not duplicate the process indication.


### TODO:
- [ ] Open the TR PP by tapping on the Live Activity widget from the locked screen (#9482)
- [x] Match the style of the  Live Activity widget with the TR PP (icons etc) (ios/add-icons-to-the-live-activity-widget)
- [x] Remove the core changes to the separate PR
- [x] Remove the ElevationViewController changes to the separate PR (#10049)

### Design:
https://www.figma.com/design/5R5LmdM6L4GyfWjSrtlFqZ/OM-Track-Recoding-by-KK?node-id=0-1&t=iHgobiqKnwgxAeCj-1

### Results:
(the icons will be gray - pls see the #10049)
![Simulator Screen Recording - iPhone 16 Pro - 2025-01-14 at 11 47 20](https://github.com/user-attachments/assets/0071698f-2786-4ab2-af21-a370400c179d)
![Simulator Screen Recording - iPhone 16 Pro - 2025-01-14 at 12 35 30](https://github.com/user-attachments/assets/a33ee773-d8b1-423a-b55e-afa0633f26a1)
![Simulator Screen Recording - iPhone 16 Pro - 2025-01-14 at 12 46 43](https://github.com/user-attachments/assets/e29531d2-7bf3-438c-ba26-4ce8b7f2c9b6)

pastk commented:
A tap on the TR widget opens the PP which is great.

What happens if its tapped again? Toggling the PP seems like a good idea.

Alternatively, when the widget is tapped it should open the PP and disappear itself. Closing the PP would make the widget appear again.
The logic here is that the PP and the widget duplicate each other.  Its going to be especially true when we expand the widget with basic stats display.
Or maybe a tap on an expanded widget would make it collapsed (just an icon; and PP would open of course)?

kirylkaveryn commented:
``

kavikhalique commented:
> Here is how it will look (selection of the other POI will enable the indicator to notify about the recording too):

IMO, changing visibility of indicator/widget while track recording is turned on would be little bit confusing for users because icon disappears also when track recording stops which may confuse users that by clicking on icon track recording has been stopped.

I think it would be better to keep the widget visible while TR is turned on even when PP opens, and while PP for TR is open, clicking outside anywhere the PP should minimize the PP and no click should be performed on map.

kirylkaveryn commented:
> IMO, changing visibility of indicator/widget while track recording is turned on would be little bit confusing for users because icon disappears also when track recording stops which may confuse users that by clicking on icon track recording has been stopped.

Valid notice!

pastk commented:
> while PP for TR is open, clicking outside anywhere the PP should minimize the PP and no click should be performed on map.

If it behaves differently from other PPs (one can tap on the map to immediately switch to another PP) then it'd be confusing.  And I see no reason to make it different either..

biodranik commented:
Thanks! Rebased and launched [alpha build](https://github.com/organicmaps/organicmaps/actions/runs/15146928187) for testflight to see it in action!

kavikhalique commented:
My UX suggestion for TR screen widget/button-

1. After starting track recording the TR widget/button should get enabled and shown on the screen.

2. On tap it should open/close the place page for current TR (i.e it will not disappear during the TR is turned on)

3. On long tap it should stop track recording, show dialog for save/discard and don't show the pp after saving.

4. There should be an alternative save button (which is already present) in the PP which will directly save the track and don't prompt user to discard or cancel and also after saving, the current pp will get converted into track pp.

5. When later pause feature will be implemented long tap on screen TR widget/button can be converted to play/pause from stop.

WDYT @biodranik @kirylkaveryn @vng

biodranik commented:
@kirylkaveryn it would be great to add some screens for non-track recording PP too. Btw, can any track PP be implemented/merged independently from recording track PP? Does it make sense?

### PP of the existing track

1. Tap on track, Tap Edit, tap Delete => PP stays (error). Suggestion: remove Edit Track button from its PP, and edit track title, its set, and its color directly by tapping on corresponding labels (color label should be added too, under X and share icons). One less click for frequent actions => good! We likely implement it for point PPs too. Need to hint a user that it's editable (maybe some visual hinting until user clicks it once?). Removing Edit button will free more space for the graph and stats.
2. Delete button can be moved to track's PP, it is not a very frequent action (or can stay on PP, let's discuss, see below).
3. Nit: show speed and the time from the track's start for the selected point.

I see several options for the bottom button bar:
1. Duplicate existing point PP (all the same buttons, the same behavior). UX complications may come from displaying both bookmark name/set and track name/set on the same PP (and editing them easily).
2. The same as (1), but pressing on a bookmark button will close track's PP and open point PP.
4. Add a button to close track PP and open point PP. Route To/Route From doesn't look like frequent actions for a point on a track. Switching PP will show them.


### Recording widget

Proposal is to move the top right recording button that appears after recording started, to the bottom button bar (5th button on the toolbar). This button is always visible and:
- starts recording if it is stopped
- shows distance and time of the currently recorded track (making a button wider for better readability is also an option)
- shows Recording Track PP on tap (buttons layout may depend on the previously discussed options above)

### Recording Track PP

- Current speed and elevation should also be visible (also in the Live Widget)
- Share button is important, even if track is not saved yet

### Pause and resume

Starting track recording should always ask "Start new track" or "Continue last track". This is a simple yet convenient UX for the pause functionality. Continue will record a separate segment for the track. Is it already supported in the core? Rendering segments separately will also be helpful later.

@kirylkaveryn @kavikhalique WDYT? Let's make a list of separate issues/steps to proceed.


P.S. @kirylkaveryn did you try to simulate gpx track locations with altitude using [approach from this article](https://medium.com/@merlos/how-to-simulate-locations-in-xcode-b0f7f16e126d)?

vng commented:
Confusing for me that I can Stop recording only by tapping the Save button in PP.

biodranik commented:
That was my concern too. With an easy way to resume recording it is not an issue )

For clarity, it can be named "Stop recording and save"

biodranik commented on iphone/CoreApi/CoreApi/PlacePageData/Common/PlacePagePreviewData.mm:
```diff
@@ -46,6 +50,16 @@ static PlacePageDataSchedule convertOpeningHours(std::string_view rawOH)
+- (instancetype)initWithTrackInfo:(TrackInfo * _Nonnull)trackInfo {
+  self = [super init];
+  if (self) {
+    // TODO: (KK) Replace separator with a shared static constant.
```
nit: This todo won't improve anything in terms of speed/reducing allocations and can be safely removed.

biodranik commented on iphone/CoreApi/CoreApi/PlacePageData/ElevationProfile/ElevationProfileData.mm:
```diff
@@ -30,23 +30,36 @@ - (instancetype)initWithTrackId:(MWMTrackID)trackId
+                                                                                  andAltitude:point.m_point.GetAltitude()];
+    [pointsArray addObject:elevationPoint];
+  }
+  return [pointsArray copy];
```
Why is copying needed?

kirylkaveryn replied:
Because the `NSMutableArray` will not be mapped to the `NSArray` automatically. 
It is a bad pattern in objc to return a mutable object if the type is declared as immutable because the caller of the method expects that the object will not be changed.

kirylkaveryn replied:
When the caller (A) of the method gets the pointer to the `NSArray`, it expects that the array behaves as immutable - no elements can be added/removed.
But the creator of the arrary (B) still can add/remove the elements in the NSMutableArray and it may unexpectedly affect the A.
`NSArray` doesn't allow adding/removing objects. So it is strange that object-creator (B) can mutate the object declared as immutable after returning it from the method.

biodranik commented on iphone/CoreApi/CoreApi/PlacePageData/ElevationProfile/ElevationProfileData.mm:
```diff
@@ -30,23 +30,36 @@ - (instancetype)initWithTrackId:(MWMTrackID)trackId
+  for (auto const & point : points) {
+    auto pointLatLon = mercator::ToLatLon(point.m_point.GetPoint());
+    CLLocationCoordinate2D coordinates = CLLocationCoordinate2DMake(pointLatLon.m_lat, pointLatLon.m_lon);
+    ElevationHeightPoint * elevationPoint = [[ElevationHeightPoint alloc] initWithCoordinates:coordinates
```
How coordinates are used in UI? Can we avoid this unnecessary transformation/allocation/copying/storage? Looks like graph uses only distance and altitude:
```
    self.chartValues = points.map { ChartValue(xValues: $0.distance, y: $0.altitude) }
    self.distances = points.map { $0.distance }

```

kirylkaveryn replied:
`coordinates` are used to draw the point on the track in Core. 
The previous core's implementation (that I removed) was to find the target point in the track points array by distance - this logic was quite complex. When the user drags the point on the chart this algorithm was exetuted hundred times per sec only to draw the point on the map. So it easier just to strore the coordinates temporarly and pass them to the core.

kirylkaveryn replied:
It is used here (last line):
https://github.com/organicmaps/organicmaps/blob/62adacf74fc152551f438d620fe88ee67ae9b12f/iphone/Maps/UI/PlacePage/Components/ElevationProfile/ElevationProfilePresenter.swift#L102-L106

biodranik commented on iphone/Maps/Classes/CustomViews/MapViewControls/TrackRecordingButtonViewController.swift:
```diff
@@ -38,31 +39,29 @@ final class TrackRecordingViewController: MWMViewController {
-    })
+  override func viewDidLoad() {
+    super.viewDidLoad()
+    // async is for smoother appearance
```
It looks like a hack, why exactly is it needed/why the appearance is not smooth? Is it because of too much work on main thread? Is it still an issue with avoiding unnecessary calculations?

biodranik commented on iphone/Maps/Classes/CustomViews/MapViewControls/TrackRecordingButtonViewController.swift:
```diff
@@ -38,31 +39,29 @@ final class TrackRecordingViewController: MWMViewController {
+  override func viewDidLoad() {
+    super.viewDidLoad()
+    // async is for smoother appearance
+    DispatchQueue.main.asyncAfter(deadline: .now() + kDefaultAnimationDuration) {
+      self.setState(self.state, completion: nil)
+    }
```
nit: Docs say:
> Specifying the current time is less efficient than calling the [async(execute:)](https://developer.apple.com/documentation/dispatch/dispatchqueue/async(execute:)) method directly.

Will this work?

```suggestion
    DispatchQueue.main.async {
      self.setState(self.state, completion: nil)
    }
```

kirylkaveryn replied:
removed!

kirylkaveryn commented:
> For clarity, it can be named "Stop recording and save"

Such long names are not suitable for the tab bar titles. On some languages it will take up to half of the screen.

I understand the disadvantages of my approach, but the "continue previous" may be distractive too - additional tap on every new recording.

biodranik commented on iphone/Maps/Core/TrackRecorder/TrackRecordingManager.swift:
```diff
@@ -84,22 +112,42 @@ final class TrackRecordingManager: NSObject {
+        try startRecording()
+      case .stopAndSave(let name):
+        stopRecording()
+        try checkIsTrackNotEmpty()
```
What's the point in using try if the catch is below? Why not a simple if check?

kirylkaveryn replied:
Maybe it is wrong solution -  the `track is empty` it a common state and should be handeled as if check. Thanks. I'll fix it

biodranik commented on iphone/Maps/Core/TrackRecorder/TrackRecordingManager.swift:
```diff
@@ -113,104 +161,85 @@ final class TrackRecordingManager: NSObject {
+      trackRecorder.startTrackRecording()
+      notifyObservers()
+      do {
+        try activityManager?.start(with: trackRecordingInfo)
```
Why can it throw an exception?

kirylkaveryn commented:
> Confusing for me that I can Stop recording only by tapping the Save button in PP.

I removed this Stop button because in the previous solution with Ann Alert I lost several tracks by accidentally tapping to the wrong button - the are look the same. I found (for me) that it is a common pattern to stop track at the end of the hike/drive when the user is tired and it can make a mistake.

Maybe to add a Cancel button with an additional confirmation (or without) will be a better idea. But to delete track is quite easy.

biodranik commented on iphone/Maps/Core/TrackRecorder/TrackRecordingManager.swift:
```diff
@@ -84,22 +112,42 @@ final class TrackRecordingManager: NSObject {
     }
   }
 
+  private func checkIsTrackNotEmpty() throws(TrackRecordingError) {
```
Is this function really needed?

kirylkaveryn replied:
I can simplify it to non thwrowing

biodranik commented on iphone/Maps/Core/TrackRecorder/TrackRecordingManager.swift:
```diff
@@ -113,104 +161,85 @@ final class TrackRecordingManager: NSObject {
-    alert.addAction(UIAlertAction(title: L("stop_without_saving"), style: .default, handler: { _ in onStop() }))
-    alert.addAction(UIAlertAction(title: L("save"), style: .cancel, handler: { _ in onSave() }))
-    UIViewController.topViewController().present(alert, animated: true)
+    DispatchQueue.main.async {
```
Why is dispatching needed?

kirylkaveryn replied:
Usually to guarantee an execution on the main thread

biodranik reviewed (commented):
Let's test it over the weekend.

biodranik commented:
Launched this PR (rebased on master) in sim with moving location and got this assert
<img width="1415" alt="выява" src="https://github.com/user-attachments/assets/73fd6d67-f046-4c8e-a3d7-b2f8d152eead" />

biodranik commented:
Simulator doesn't show the graph
![выява](https://github.com/user-attachments/assets/65d185ad-7681-4f5d-8e37-1471edf062ab)

biodranik commented:
Looks like it reproduces because locations in the simulator do not have elevation…

biodranik commented:
1. When recording starts, the PP is not fully open because the graph is empty. When graph fills with data, PP is not automatically expanded (and user doesn't know that there's a graph in it). It may be better to show a fully open PP with an empty graph, and explicitly state there "No data" or something similar.
2. There is a bug with the graph (dot is not on the line, see screenshot and attachment below).
3. I've added mock elevation into simulator's location, it may be easier to test with it: https://github.com/organicmaps/organicmaps/pull/10682

[9 чэрвеня 2025 г. у 0009.kmz.zip](https://github.com/user-attachments/files/20647042/9.2025.0009.kmz.zip)

![выява](https://github.com/user-attachments/assets/3fa719a1-7405-42b1-9e9e-deeba4876d1d)

kirylkaveryn commented:
> * When recording starts, the PP is not fully open because the graph is empty. When graph fills with data, PP is not automatically expanded (and user doesn't know that there's a graph in it). It may be better to show a fully open PP with an empty graph, and explicitly state there "No data" or something similar.

Fixed! Now the screen is fully opened with an empty ele chart.

> * There is a bug with the graph (dot is not on the line, see screenshot and attachment below).

The bug is not related to this feature and this track example is a bit synthetic. 
The issue created https://github.com/organicmaps/organicmaps/issues/10690

> * I've added mock elevation into simulator's location, it may be easier to test with it: [[ios] Add random smoothly changed elevation to iOS Simulator for testing track recording #10682](https://github.com/organicmaps/organicmaps/pull/10682)

Thanks! It really helps.

kirylkaveryn commented on iphone/CoreApi/CoreApi/PlacePageData/Common/PlacePagePreviewData.h:
```diff
@@ -39,6 +40,8 @@ NS_ASSUME_NONNULL_BEGIN
 @property(nonatomic, readonly) PlacePageDataSchedule schedule;
 @property(nonatomic, readonly) BOOL isMyPosition;
 
+- (instancetype)initWithTrackInfo:(TrackInfo * _Nonnull)trackInfo;
```
Usually the nonnull and nullable use it as it property attribute and the __Nonull and __Nullable I use it as a key works for work variables and methods arguments
